### PR TITLE
BBS-140: Load materials from Primo

### PR DESCRIPTION
**Pull request by kasperg**
This PR contains the first stab at loading materials from Primo.

It consists of the following parts:

- A stub implementation of the Primo `search` provider. This will bounce calls back to OpenSearch if they are not implemented by Primo
- A proper client library for communicating with the Primo Brief Search webservice
- An implementation of the `search_object_load` provider hook and initial implementation of `TingObjectInterface` for Primo
- Logging of requests
- Proper retrival for covers

### CODE REVIEW

**Comment by kasperg** [comment]
Scrutinizer complains about indentation. I do not know why. I think it can be disregarded.

File: modules/primo/includes/primo.search.inc [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+/**
+ * @file
+ * Implements searching in Primo
+ */
+
+use Primo\BriefSearch\Document;
+use Primo\Ting\Object;
+use Ting\Search\TingSearchRequest;
+
+// TODO Remove this once Primo search provider is fully implemented.
```
Comment (tripox) [inline modules/primo/includes/primo.search.inc:12]: Maybe "TODO Primo" here, so it's easier to find later.

File: modules/primo/primo.module [hunk 2]
```diff
@@ -87,6 +93,36 @@ function primo_settings_form() {
...
+    $handler->push($logger, 'logger');
+  }
+
+  return new Client(
+    new GuzzleClient([
+      'handler' => $handler,
+      'base_uri' => variable_get('primo_base_url')
+    ]),
+    variable_get('primo_institution_code'),
+    ip_address()
```
Comment (tripox) [inline modules/primo/primo.module:121]: Should this be the IP of the client machine? For logging?

File: modules/primo/src/Primo/BriefSearch/Client.php [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+   * @return \Primo\BriefSearch\Result
+   *   The search result.
+   *
+   * @throws \Primo\Exception\TransferException
+   *   If an error occurs during the execution of the search.
+   */
+  public function search($query, $offset, $numResults, $parameters = []) {
+    $parameters += [
+      'query' => $query,
+      'indx' => $offset,
```
Comment (tripox) [inline modules/primo/src/Primo/BriefSearch/Client.php:109]: index?

File: modules/primo/src/Primo/BriefSearch/Client.php [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+   * @return bool
+   *   Whether the url is a valid thumbnail url.
+   */
+  public function validateThumbnail($url) {
+    // Use HEAD as we only care about status code and content type here.
+    $response = $this->httpClient->head($url);
+    // According to docs content type should be a string but is really an array.
+    /* @var string[] $contentType */
+    $contentType = $response->getHeader('Content-Type');
+    return $response->getStatusCode() == 200 &&
```
Comment (tripox) [inline modules/primo/src/Primo/BriefSearch/Client.php]: We could safely use `===` here.

File: modules/primo/src/Primo/BriefSearch/Document.php [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+   */
+  public function getYear() {
+    $nodes = $this->recordXpath($this->recordId, '//primo:addata/primo:date');
+    return $this->nodeValue($nodes);
+  }
+
+  /**
+   * Gets the value of a local display field.
+   *
+   * Primo supports local fields which are represented as XML elements namd
```
Comment (tripox) [inline modules/primo/src/Primo/BriefSearch/Document.php]: named*

File: modules/primo/src/Primo/BriefSearch/Result.php [hunk 6]
```diff
@@ -0,0 +1,40 @@
+<?php
+
+namespace Primo\BriefSearch;
+
+/**
+ * A result from accesing the Primo brief search servide.
```
Comment (tripox) [inline modules/primo/src/Primo/BriefSearch/Result.php]: service*

File: modules/primo/src/Primo/Ting/Object.php [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+  public function getSeriesDescription() {
+    // TODO: Implement getSeriesDescription() method.
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function getSource() {
+    $source = $this->document->getLocalDisplayField(6);
+    return str_replace('Á frummáli: ', '', $source);
```
Comment (tripox) [inline modules/primo/src/Primo/Ting/Object.php]: I guess `Á frummáli: ` is something in Icelandic indicating a text related to showing the source.
A comment would still be nice.

File: modules/primo/primo.module (hunk 2 above)
Comment (tripox) [reply modules/primo/primo.module:121]: Ah, that was the client IP. :)

**Review by tripox — CHANGES_REQUESTED** [review]
A few comments, mostly typo's. 
One of them is in a key, so I think that one should be fixed at least (unless it's not a typo).
Otherwise it looks very good to me. 👍

File: modules/primo/primo.module (hunk 2 above)
Comment (kasperg) [reply modules/primo/primo.module:121]: :+1:

File: modules/primo/src/Primo/BriefSearch/Client.php (hunk 3 above)
Comment (kasperg) [reply modules/primo/src/Primo/BriefSearch/Client.php:109]: No - [the parameter is actually called `indx`](https://developers.exlibrisgroup.com/primo/apis/webservices/xservices/search/briefsearch) for some reason.

Comment (tripox) [reply modules/primo/src/Primo/BriefSearch/Client.php:109]: Okay, I'll approve then. :)

**Review by tripox — APPROVED** [review]

**Pull request merged**